core/primitives/src/utils.rs

```diff
@@ -469,23 +471,55 @@ where
     Serializable(object)
 }
 
-/// Derives `AccountId` from `PublicKey``.
+// TODO(eth-implicit) Replace this function (and wat dependency) with real Wallet Contract implementation.
+pub fn wallet_contract_placeholder() -> ContractCode {
+    let code = wat::parse_str(r#"(module (func (export "main")))"#);
+    ContractCode::new(code.unwrap().to_vec(), None)
+}
+
+/// From `near-account-id` version `1.0.0-alpha.2`, `is_implicit` returns true for ETH-implicit accounts.
+/// This function is a wrapper for `is_implicit` method so that we can easily differentiate its behavior
+/// based on the protocol version.
+pub fn account_is_implicit(account_id: &AccountId, protocol_version: ProtocolVersion) -> bool {
+    if checked_feature!("stable", EthImplicit, protocol_version) {
+        account_id.get_account_type().is_implicit()
+    } else {
+        account_id.get_account_type() == AccountType::NearImplicitAccount
+    }
+}
+
+/// Derives `AccountId` from `PublicKey`.
 /// If the key type is ED25519, returns hex-encoded copy of the key.
-pub fn derive_near_implicit_account_id(public_key: &ED25519PublicKey) -> AccountId {
-    hex::encode(public_key).parse().unwrap()
+/// If the key type is SECP256K1, returns '0x' + keccak256(public_key)[12:32].hex().
+pub fn derive_account_id_from_public_key(public_key: &PublicKey) -> AccountId {
+    match public_key.key_type() {
+        KeyType::ED25519 => hex::encode(public_key.key_data()).parse().unwrap(),
+        KeyType::SECP256K1 => {
```

Comment:
nit: I slightly dislike that we rely on an implementation detail of each account encryption scheme in order to differentiate between them. Would it be possible to supply a hint saying what type of account is expected and error out if it doesn't match the public key? 

It's a small nit, please only do this if you agree, the code is simple and decent enough.

Reply:
I agree that it is slightly misleading. I would split it into two functions `derive_near_implicit_account_id` and `derive_eth_implicit_account_id` and let the compiler enforce that `ED25519` and `SECP256K1` are passed respectively.